### PR TITLE
JVM IR: Move direct invoke optimization into a separate pass and fix tests

### PR DESCRIPTION
**Pull request by sfs**
The direct invoke optimization broke a lot of our callable reference tests (which didn't end up generating any callable references!) and resulted in some problems for the Kotlin ABI by generating unnamed local functions, leading to `lambda-n` methods in the compiled program. This is also a problem for the inliner, which is why the optimization was disabled inside of inline functions.

This PR fixes the test cases and changes the direct invoke optimization to inline lambdas instead of generating calls to local functions. This does lead to different stepping behavior in the debugger, but at the moment there is no good way to fix this, since the IR does not have a mechanism to encode the relevant debug information (explicit line numbers and inline frames).

### CODE REVIEW

**Comment by max-kammerer** [comment]
>since the IR does not have a mechanism to encode the relevant debug information (explicit line numbers and inline frames).

NB: @ivandev0 probably could have some prototype for that supporting IR-inliner in JVM backend

**Comment by sfs** [comment]
That would be great, since we currently run into the same issue with lambdas in array initializers and assert messages. These are the other two places where we already use IR inlining in the JVM backend.

**Comment by udalov** [comment]
@sfs Can you please report a KT issue with examples of what's broken? We need an issue for any major problem or change, so that everyone in the team (including QA, release team, sometimes tech writers) can understand what the problem is and when it is fixed.

**Comment by sfs** [comment]
There shouldn't be anything that's broken externally. It was our callable reference tests that weren't actually generating callable references. The previous direct invoke optimization also excluded all invokes in the scope of inline functions, because that wouldn't have worked.

So there's three things I could write issues for.
- The lack of direct invoke optimization in inline functions,
- a feature request for different codegen around direct invokes,
- and/or a debug experience bug for the behavior *after* this PR is merged?

Not sure if any of these would be all that interesting for external users, but let me know if you thing that would be useful.

**Comment by udalov** [comment]
@sfs OK, I see, thanks for the explanation! Indeed, not a major change for the users, so it's not necessary then.

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/ir/IrInlineUtils.kt [hunk 1]
```diff
@@ -30,12 +30,7 @@ class IrInvokable(val invokable: IrValueDeclaration) : IrInlinable()
 class IrInlinableLambda(val function: IrSimpleFunction, val boundReceiver: IrValueDeclaration?) : IrInlinable()
 
 // Return the underlying function for a lambda argument without bound or default parameters or varargs.
-private fun IrExpression.asInlinableLambda(builder: IrStatementsBuilder<*>): IrInlinableLambda? {
-    if (this is IrFunctionExpression) {
-        if (function.valueParameters.any { it.isVararg || it.defaultValue != null })
-            return null
-        return IrInlinableLambda(function, null)
```
[the remark this answers is not in the document]
Comment (sfs) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/ir/IrInlineUtils.kt:37]: I completely missed this. Removed the commit and kept the code in `asInlinableLambda`.

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/FunctionReferenceLowering.kt [hunk 2]
```diff
@@ -103,108 +100,6 @@ internal class FunctionReferenceLowering(private val context: JvmBackendContext)
...
-        return newBlock
-    }
-
-    private fun rewriteDirectInvokeToFunctionReference(irInvokeCall: IrCall, irFunRef: IrFunctionReference): IrExpression? {
-        // TODO deal with type parameters somehow?
-        // It seems we can't encounter them in the code written by user,
-        // but this might be important later if we actually perform inlining and optimizations on IR.
-        return when (val irFun = irFunRef.symbol.owner) {
-            is IrSimpleFunction -> {
-                if (irFun.typeParameters.isNotEmpty()) return null
```
[the remark this answers is not in the document]
Comment (sfs) [reply compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/FunctionReferenceLowering.kt:164]: Good catch, I had moved the comment and check to `DirectInvokeLowering.visitFunctionReference` and later moved the check to `DirectInvokeLowering.visitCall`. I've moved the comment there as well now.

**Comment by udalov** [comment]
Merged manually with a few fixes in tests in 3b613b3745c. Thanks!

**Comment by udalov** [comment]
Reverted the main commit in b50d2ff20a6 because it leads to [KT-53202](https://youtrack.jetbrains.com/issue/KT-53202)

**Pull request closed without merging**